### PR TITLE
tests: on_target: add memfault test

### PR DESCRIPTION
**Pull request by dilner**

### CODE REVIEW

File: tests/on_target/tests/test_memfault.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import pytest
```
Comment (jorgenmk) [inline tests/on_target/tests/test_memfault.py:1]: License header?

File: tests/on_target/tests/test_memfault.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    t91x_board.uart.xfactoryreset()
+    patterns_memfault = [
+            "Network connectivity established",
+            "memfault: memfault_task: Memfault module task started",
+            "memfault: memfault_task: Cloud status received",
+    ]
+
+    # Save timestamp of latest heartbeat event
+    heartbeat_events = get_latest_events("HEARTBEAT", IMEI)
+    timestamp_old_reboot_evt = timestamp(heartbeat_events[0]) if heartbeat_events else  None
```
Comment (jorgenmk) [inline tests/on_target/tests/test_memfault.py:75]: `timestamp_old_*heartbeat*_evt` ?

File: tests/on_target/tests/test_memfault.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    t91x_board.uart.flush()
+    reset_device()
+    t91x_board.uart.wait_for_str(patterns_memfault, timeout=60)
+
+    # Trigger hard fault to generate memfault event
```
Comment (jorgenmk) [inline tests/on_target/tests/test_memfault.py:81]: `# Trigger bus fault...` ?

File: tests/on_target/tests/test_memfault.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+def wait_for_heartbeat(timestamp_old_reboot_evt):
+    new_heartbeat_found = False
+    start = time.time()
+    while time.time() - start < MEMFAULT_TIMEOUT:
+        logger.debug("Looking for latest heartbeat events")
+        time.sleep(5)
+        new_heartbeat_events = get_latest_events("HEARTBEAT", IMEI)
+        if not new_heartbeat_events:
+            continue
+        if not new_heartbeat_events[0]["summary"]["event_info"]["metrics"]["MemfaultSdkMetric_UnexpectedRebootCount"] == 1:
```
Comment (jorgenmk) [inline tests/on_target/tests/test_memfault.py:50]: Perhaps this should be wrapped in a `try/except (KeyError, ValueError, ...): continue` statement?

Can this `UnexpectedRebootCount` increase past 1 when the test is run multiple times?

File: tests/on_target/tests/test_memfault.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            continue
+        # Check that we have heartbeat event with newer timestamp
+        if not timestamp_old_reboot_evt:
+            new_heartbeat_found = True
+        elif timestamp(new_heartbeat_events[0]) > timestamp_old_reboot_evt and not new_heartbeat_found:
+            new_heartbeat_found = True
+        if new_heartbeat_found:
+            break
+    else:
+        raise RuntimeError("No new heartbeat event observed")
```
Comment (jorgenmk) [inline tests/on_target/tests/test_memfault.py:60]: I think this should be `AssertionError` ?

File: tests/on_target/tests/test_memfault.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
+import pytest
+import os
+import requests
+import datetime
+import time
+from utils.flash_tools import flash_device, reset_device
+import sys
+sys.path.append(os.getcwd())
```
Comment (jorgenmk) [inline tests/on_target/tests/test_memfault.py:8]: I think this can be removed? We are importing from `utils` two lines up.

**Review by DematteisGiacomo — APPROVED** [review]

**Review by jorgenmk — APPROVED** [review]

**Pull request merged**